runbotics/runbotics-scheduler/src/scheduler-database/credential-collection/credential-collection.service.ts

```diff
@@ -81,56 +82,82 @@ export class CredentialCollectionService {
         }
 
         if (sharedWith && sharedWith.length > 0) {
-            const credentialCollectionUserArray =
-            await this.getCollectionUserArrayWithPrivileges(
-                    sharedWith,
-                    credentialCollection,
-                    user,
-                    tenantId,
-                );
+            const credentialCollectionUserArrayToSave =
+                await this.getCollectionUserArrayWithPrivileges(
+                        sharedWith,
+                        credentialCollection,
+                        userDto,
+                        tenantId,
+                    );
+
+            credentialCollection.credentialCollectionUser =
+                credentialCollectionUserArrayToSave;
+        }
 
-                credentialCollection.credentialCollectionUser =
-                credentialCollectionUserArray;
-            }
         return this.credentialCollectionRepository.save(credentialCollection);
     }
 
-    async findAllAccessible(user: IUser) {
-        const collections = await this.credentialCollectionRepository.find({
-            where: {
-                tenantId: user.tenantId,
-                credentialCollectionUser: {
-                    userId: user.id,
-                    privilegeType: PrivilegeType.WRITE,
-                },
-            },
-            relations,
-        });
-
-        if (collections.length === 0) {
-            throw new NotFoundException('Could not find any credential collections');
-        }
-
-        return collections;
+    findAllAccessible(user: IUser) {
+        return this.credentialCollectionRepository
+            .createQueryBuilder('credentialCollectionEntity')
+            .innerJoinAndSelect('credentialCollectionEntity.createdBy', 'createdBy')
+            .innerJoin(
+                'credentialCollectionEntity.credentialCollectionUser',
+                'filteredCredentialCollectionUsers',
+            )
+            .where(
+                'credentialCollectionEntity.tenantId = :tenantId',
+                { tenantId: user.tenantId },
+            )
+            .andWhere(
+                'filteredCredentialCollectionUsers.user.id = :userId',
+                { userId: user.id },
+            )
+            .andWhere(
+                'filteredCredentialCollectionUsers.privilegeType = :privilegeType',
+                { privilegeType: PrivilegeType.WRITE },
+            )
+            .innerJoinAndSelect(
+                'credentialCollectionEntity.credentialCollectionUser',
+                'allCredentialCollectionUsers',
+            )
+            .innerJoinAndSelect('allCredentialCollectionUsers.user', 'allUsers')
+            .getMany();
     }
```

Comment:
well congrats for achieving that in one query

but it's a huge change in the readability
1. You can merge these 3 'where' conditions into one - here example of using 2 conditions in one 'where' https://typeorm.io/select-query-builder#getting-values-using-querybuilder
2. I'm quite afraid for seeing so many inner joins in method that is going to be called very often for checking if user has access to some credential. But let's put off this problem for later when it'll arise.

Reply:
Do you really need to check createdBy? I think you're automatically add credential owner to credentialCollectionUser relation

Reply:
I do not check `createdBy`, I retrieve relationships with this property